docs/proposals/dns-using-envoy.md

```diff
@@ -0,0 +1,69 @@
+# Use Envoy DNSFilter for DNS resolution
+
+## Context
+
+Currently, Kuma relies on an external DNS to provide VIPs which are then added as services which proxy back to the actual Cluster in Envoy.
+For VIPs to be resolved Kuma CP runs a DNS server which maintains this mapping.
+
+This current setup has some limitations:
+
+- Apps need to set their DNS to point to the CP's DNS (or CNIs need to configure DNS to be correct)
+- The CP needs to run a DNS and handle the load
+- When the CP is down DNS requests will fail. This is usually not the case in a Service Mesh
+- DNS configuration is global to all meshes
+- DNS is a separate component which makes it less customisable compared to things that end up in XDS
+- This DNS is not aware of the origin of requests, therefore isolation of meshes breaks as anyone can resolve any domain
+
+In this proposal we suggest using [Envoy's DNSFilter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/udp_filters/dns_filter) to replace the CP DNS.
+We add a new section in the Mesh Resource to be able to control the configuration of the DNS.
+
+## Out of scope
+
+* UDP transparent proxying
+* Handling SRV records. While DNSFilter supports it, it's probably not necessary in v1
+
+## Requirements
+
+* The top-level domain should be configurable (defaults to `mesh`)
+* Drop in replacement to Kuma DNS (Default configuration should match Kuma's current configuration)
+* ability to set upstream dns resolvers to have a single resolver in the configuration
+* Select the tag(s) to use for lower level domain names
+* Provide complete isolation between meshes. .i.e. it shouldn't be possible from a mesh to figure out services running in other meshes by querying the DNS
+* DNS query should still succeed even if the CP is down
+
+## Configuration model
+
+Here is a full example of the configuration on the mesh.
+
+```yaml
+type: Mesh
+name: default 
+dns:
+  domain: "kuma" # [Optional defaults to "mesh"] the TLD to use for services exposed with the DNS
+  upstreamResolvers: # [Optional] A list of resolvers to use when we resolving names not in the TLD
+    - "8.8.8.8"
+    - "4.4.4.4"
+  address: "169.254.53.53" # [Optional defaults to 0.0.0.0] the address on which the UDP listener should bind in envoy
+  port: "53" # [Optional defaults to 53 or redirectOutbound if transparent proxying is set] the port on which the UDP listener should bind in envoy
```

Comment:
Because redirectOutbound is currently only TCP we can add UDP without problem (DNSFilter only works on UDP DNS).

Reply:
> Also if we are picking port, there probably should be an option to override this for one dataplane. Just like we do with prometheus endpoint.

I guess that would be an extra config in the dataplane spec?

Reply:
> Because redirectOutbound is currently only TCP we can add UDP without problem (DNSFilter only works on UDP DNS).

What is the problem that prevents us from binding to any other port? Is it about iptables? I thought we are touching only TCP traffic with iptables.

> I guess that would be an extra config in the dataplane spec?

correct

Reply:
> What is the problem that prevents us from binding to any other port? Is it about iptables? I thought we are touching only TCP traffic with iptables.

If by any you mean any arbitrary port, nothing prevents you from doing this. Would you want to always default to 53 regardless or transparent proxying or not?